tensorflow/compiler/xla/service/gpu/BUILD

```diff
@@ -1357,19 +1356,25 @@ tf_cc_test(
 
 cc_library(
     name = "buffer_comparator",
-    srcs = ["buffer_comparator.cc"],
+    srcs = [] + if_cuda_is_configured([
+        "buffer_comparator_cuda.cc",
+    ]) + if_rocm_is_configured([
```

Comment:
This is also a compile-time dispatch that result in similar problems from `#ifdef`. Can you make buffer_comparator a runtime dispatch similar to what you did with redzone allocator?

Reply:
Thanks for the review. Let me try to understand the right way to do it:

The way I did with redzone allocator in the build file is that:
```
cc_library(
     name = "redzone_allocator",
     srcs = [] + if_rocm_is_configured([
         "//tensorflow/stream_executor/rocm:redzone_allocator.cc",
     ]) + if_cuda_is_configured([
         "//tensorflow/stream_executor/cuda:redzone_allocator.cc",
     ]),
     hdrs = ["redzone_allocator.h"],
     deps = [
     ...
```
It seems to be the exact same approach as in here, except that the component of redzone allocator is under stream executor (which I assume is the reason why it is not causing a problem from your end). 

Hip runtime compilation feature is not ready for us just yet, which is why I have been carefully not to implement related Cuda side runtime ptx compilation features, eg, `ptxas_utils.cc::CompilePtxOrGetCached()`. If the goal is to get rid of compile-time dispatch like `if_rocm_is_configured` from XLA, I can bring that part of interface parallel to ROCm as well.

Reply:
#ifdef'ing the whole is worse than switching source at build time, which is worse than an isolated #ifdef, which is worse than virtual or runtime dispatch.
Hence the question: can we avoid build-time selection entirely? I understand that this would be a non-trivial change.

Reply:
Agreed on the sequence of ugliness :-). The rest of the thoughts sounds very ambitious if I understand it right. Based on the chain, switching the source at build time is worse than an isolated `#ifdef`. On a previous conversation I had with @timshen91, the impression I have is that the aim for XLA compilation is to compile without `--confg=rocm or cuda`. This makes us not having the option of doing any kind of `#ifdef`, and leaving the only option to runtime dispatch. (Let me know if there's any other way to do scoped `#ifdef`, independent from `--config` approach).

Following runtime dispatch thought, since `Cuda` and `ROCm` relying on different run-time libraries. To be able to get rid of build-time selection means that those run-time libraries has to exist on the same machine, or on a wired scenario `Cuda` run-time needs to exist on an `AMD` machine and vice versa. I'm not sure if this is acceptable. Even worse, because the whole code base needs to compile with both compilers this implicitly requires `HIP` to be able to compile with `Cuda` run-time. 

In the specific case, if I'm to get rid of `stream_executor`'s build-time selection. I am effectively exposing the ptx compilation and its very specific `Cuda` utilities to `ROCm` side of binary. It sounds like a bad idea.

If all the above concerns does not turn out to be a problem the run-time switching is not hard to achieve: For each device, try to call either `hipinit()` or `cuinit()`. Only one can succeed, and based on if the result is successful we can tell if this is on `ROCm` or on `Cuda`.

Reply:
@jerryyin Do you want to discuss this over a call?

1) Removing #ifdefs out of TensorFlow proper is indeed a very large project which we may never finish. I'm talking about a much smaller scope: not having #ifdefs in stream executor and XLA.

> To be able to get rid of build-time selection means that those run-time libraries has to exist on the same machine

Why is this a problem? Those are just binaries we link to the final executable, right?

 >Even worse, because the whole code base needs to compile with both compilers this implicitly requires HIP to be able to compile with Cuda run-time.

StreamExecutor and XLA do not have CUDA kernels, so this should not be an issue.

> I am effectively exposing the ptx compilation and its very specific Cuda utilities to ROCm side of binary

Could you elaborate what precisely is bad? What kind of bugs do we potentially create this way?

Reply:
@cheshire Ah I've just seen your response this morning. In case we'd need further discussion, my email: zhuoran.yin@amd.com. The problem is mainly on compile time dependencies. For example, in [gpu_types.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/gpu/gpu_types.h#L45-L81), or [gpu_device.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device.cc#L87-L101). I can hardly think of a way to get rid of those macros.

Taking one step back, this discussion is out of scope of the PR, which was aimed only for adding `ROCm` parallel logic to conv algorithm picker. In an offline discussion with @whchung, we agreed that this might be a good topic bring up from TF side so that we can talk about further collaborations. We also need to review the responsibilities between each team as well, e.g. bringing up the `Cuda` side of run-time may not be `ROCm` responsibility. Before this discussion happens, I can hardly think of a way to remove compile dispatch from `redzone_allocator`.

Now back to this single issue in `buffer_comparator`'s compile time dispatch. I can try to remove it by bring `stream_executor::ptxas_util` interface to parallel.

Reply:
OK it's a good point that we have to limit the scope of this patch.

For a start, it seems easy enough to peel of the changes to RedzoneAllocator and BufferComparator?

Since those are currently no-ops for `ROCm`, it seems to me it would be best to:

 - Have a base class `RedzoneAllocator`
 - Have a method `static std::unique_ptr<RedzoneAllocator> GetAllocator()`
 - Have two subclasses: `NoopRedzoneAllocator` and `CudaRedzoneAllocator`
 - At runtime, dispatch to the second one if CUDA is available, and to the first one otherwise.
 - Perform the same steps for `BufferComparator`

@timshen91 @jerryyin Does it make sense? Any objections?

I've added quite a lot of other comments on this patch, but since it's so large it becomes hard to talk about it.
Once we limit the scope to convolutions only, it should get easier.

Reply:
@cheshire I'm planning on to start a new PR that includes only changes under stream executor. When that is done, come back to address compiler/xla related feedback.

First, I'm not sure if inheritance gives us anything more than the existing structure. Actually part of the lessen learned in exactly this PR is to avoid inheritance when possible, especially between `ROCm` and `Cuda` functionalities. This is the reason why it ended up in `PickBestAlgorithmNoCacheRocm()` and `PickBestAlgorithmNoCacheCuda()` instead of a inheritance structure. The problem with inheritance in active developing component is that when responsibility between child/parent class changed, a lot of case it will break `ROCm` build and requires constant refactoring.

Second for future extensiblility,  `ROCm` side plans on to actually implement the `RedzoneAllocator` as a mid-priority feature. From this perspective `NoopRedzoneAllocator` will be it a short-lived component.

Third, runtime dispatch is not as straightforward in the `redzone_allocator` component. The main roadblock comes from the divergence of `cuda/redzone_allocator.cc`. `cuda::redzone_allocator` take advantage of `ptxas_utils.cc`, which is dependent on `Cuda` runtime. For the minimal, the compile dispatch has to happen for `ptxas_utils`. Conceptually speaking, this is an allocator component, dispatching logic should stay outside of its responsibility.

My suggestion with the two components is to stay with the current implementation and we do not over-optimize. We can always refactor after `ROCm` side implementation for stream_executor is out, should we decide there is a lot of code duplication at that point, etc.

Reply:
> First, I'm not sure if inheritance gives us anything more than the existing structure. 

Sure, it's up to you how exactly it should be done.

> Second for future extensiblility, ROCm side plans on to actually implement the RedzoneAllocator as a mid-priority feature

That's neither here nor there, but RedzoneAllocator was written for NVidia because it is closed source and does not support sanitization. It might make less sense for AMD.

I've mentioned `NoopRedzoneAllocator/BufferComparator` because in that case we would not need to add any extra compile-time guards apart from what we have, right?

Reply:
@cheshire Want to make sure we are on the same page now.

> RedzoneAllocator was written for NVidia because it is closed source and does not support sanitization. It might make less sense for AMD.

Good information. I noticed that the usage of RedzoneAllocator has been primarily related with convolution algorithms. Is it the case where you guys experienced most cudnn bugs,  in which the pattern is that cudnn writes more than it has claimed? or is it more related with we always have multiple candidates for convolution, and can discard problematic ones in the profiling round? Knowing this will help us make better decision on the form of this check. For example, it might make better sense to be a unit test in `MIOpen` rather than in TF runtime.

>  I've mentioned NoopRedzoneAllocator/BufferComparator because in that case we would not need to add any extra compile-time guards apart from what we have.

Does the existing review count as "extra" compile-time guards? I'm trying to understand what exactly changed from the first time this PR gets approved to it goes under re-review, possibly due to compile-time dispatch in XLA component? 

For the very minimal, I will have to use compile-time dispatch in `stream_executor::ptxas_utils.cc`. With it, I will be able to get rid of the compile time dispatch from `RedzoneAllocator` and `BufferComparator`. Would that approach work?

Reply:
> Is it the case where you guys experienced most cudnn bugs, in which the pattern is that cudnn writes more than it has claimed

Reply:
> Is it the case where you guys experienced most cudnn bugs, in which the pattern is that cudnn writes more than it has claimed? 

Yes, we have seen cudnn bugs with OOB writes.

> or is it more related with we always have multiple candidates for convolution, and can discard problematic ones in the profiling round

But once a convolution has done an OOB write we can't be 100% sure that it hasn't overwritten a crucial piece of memory.

> For example, it might make better sense to be a unit test in MIOpen rather than in TF runtime.

But we don't know which convolutions are problematic on which shapes in advance.
If we have open-source kernels ideally we should try to get asan working, but it would be a much larger investment.

> I'm trying to understand what exactly changed from the first time this PR gets approved to it goes under re-review, possibly due to compile-time dispatch in XLA component?

I'm grateful to you working on this, and I apologize for the delays from our side.
I think some decisions on what is the best way to do things are still being figured out, hence the discrepancy.

For now:

 - Yes, some amount of compile-time guards is unavoidable
 - Let's minimize/localize that as much as possible
 - Hence for BufferComparator/RedzoneAllocator you don't really need this, as ROCm version does not do anything, and having a noop version might be useful in any case.

Reply:
Ok, I think we are on the same page now. Thanks for your patience. I will come up with the PR that brings `ptxas_utils` to parallel to `ROCm`, and be the only component that does compile time dispatch. Once this is done, it will naturally remove the need to have a dummy implementation of `BufferComparator` or `RedzoneAllocator`. When my new PR is out, I will ping you and @timshen91 for review, as well as leaving a new message in this PR, and we can address any new issues from there.

Reply:
PR #31908 created, as well as addressed all review feedback there. I will rebase this PR once #31908 is merged.